pkg/volumes/openstack/volumes.go

```diff
@@ -0,0 +1,337 @@
+
+type InstanceMetadata struct {
+	Name             string `json:"name"`
+	ProjectID        string `json:"project_id"`
+	AvailabilityZone string `json:"availability_zone"`
+	Hostname         string `json:"hostname"`
+	ServerID         string `json:"uuid"`
+}
+
+// OpenstackVolumes is the Volumes implementation for Openstack
+type OpenstackVolumes struct {
+	meta *InstanceMetadata
+
+	matchTagKeys []string
+	matchTags    map[string]string
+
+	computeClient *gophercloud.ServiceClient
+	volumeClient  *gophercloud.ServiceClient
+	clusterName   string
+	project       string
+	instanceName  string
+	internalIP    net.IP
+	nameTag       string
+}
+
+var _ volumes.Volumes = &OpenstackVolumes{}
+
+// NewOpenstackVolumes builds a OpenstackVolume
+func NewOpenstackVolumes(clusterName string, volumeTags []string, nameTag string) (*OpenstackVolumes, error) {
+
+	metadata, err := getLocalMetadata()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get server metadata: %v", err)
+	}
+
+	stack := &OpenstackVolumes{
+		meta: metadata,
+	}
+
+	for _, volumeTag := range volumeTags {
+		tokens := strings.SplitN(volumeTag, "=", 2)
+		if len(tokens) == 1 {
+			stack.matchTagKeys = append(stack.matchTagKeys, tokens[0])
+		} else {
+			stack.matchTags[tokens[0]] = tokens[1]
+		}
+	}
+
+	compute, err := getComputeClient()
+	if err != nil {
+		return nil, fmt.Errorf("Could not build OpenstackVolumes: %v", err)
+	}
+	stack.computeClient = compute
+
+	volume, err := getVolumeClient()
+	if err != nil {
+		return nil, fmt.Errorf("Could not build OpenstackVolumes: %v", err)
+	}
+	stack.volumeClient = volume
+
+	err = stack.discoverTags()
+	if err != nil {
+		return nil, err
+	}
+	stack.nameTag = nameTag
+
+	return stack, nil
+}
+
+func getLocalMetadata() (*InstanceMetadata, error) {
+	var meta InstanceMetadata
+	var client http.Client
+	resp, err := client.Get(MetadataLatest)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusOK {
+		bodyBytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+		err = json.Unmarshal(bodyBytes, &meta)
+		if err != nil {
+			return nil, err
+		}
+		return &meta, nil
+	}
+	return nil, err
+}
+
+func getCredential() (*gophercloud.AuthOptions, error) {
+
+	// prioritize environment config
+	env, enverr := openstack.AuthOptionsFromEnv()
+	if enverr != nil {
+		return nil, fmt.Errorf("Could not initialize openstack from environment: %v", enverr)
+	}
+	return &env, nil
+
+}
+
+func getVolumeClient() (*gophercloud.ServiceClient, error) {
+	authOption, err := getCredential()
+	if err != nil {
+		return nil, fmt.Errorf("error building openstack storage client: %v", err)
+	}
+	provider, err := openstack.NewClient(authOption.IdentityEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("error building openstack storage client: %v", err)
+	}
+	cinderClient, err := openstack.NewBlockStorageV2(provider, gophercloud.EndpointOpts{
```

Comment:
we should support cinder v3 in some point as well. Seems that this whole thing contains much hardcoded v2 stuff.